proposals/p5337.md

```diff
@@ -0,0 +1,1057 @@
+
+interface B {
+  let T:! type;
+  // Okay
+  extend impl as A where .T = T*;
+}
+
+class C {
+  impl as B {
+    // `T` here is `B.T`
+    where T = i32;
+    // `F` here is `B.F` and `A.F`.
+    fn F() -> i32*;
+  }
+}
+```
+
+This is equivalent to:
+
+```carbon
+interface A {
+  let T:! type;
+  fn F() -> T;
+}
+
+interface B {
+  let T:! type;
+  // Skip A.T due to conflict.
+  fn F() -> T*;
+}
+
+impl forall [U:! B] U as A where .T = .(B.T)* {
+  fn F() -> T = U.(B.F);
+}
+
+class C {
+  impl as B {
+    where T = i32;
+    fn F() -> i32*;
+  }
+}
+```
+
+This was used in the `OutputForwardIterator` example in
+[the "Proposal" section](#proposal).
+
+Name conflicts are not expected to generally be a problem in practice since this
+form of extension often requires the two interfaces to be defined in the same
+file.
+
+### Final `impl` priority
+
+With the design prior to this proposal, a final `impl` can only overlap another
+(final or non-final) `impl` if they have the same definition on their overlap.
+This means that if there is a `final` blanket impl of an interface, as is
+generated by `extend final impl as` in the interface's scope, then other `impl`s
+of that interface may be overly restricted. This example:
+
+```carbon
+package Core;
+
+interface As(T:! type) {
+  fn Convert[self: Self]() -> T;
+}
+
+interface ImplicitAs(T:! type) {
+  extend final impl as As(T);
+}
+```
+
+is equivalent to:
+
+```carbon
+package Core;
+
+interface As(T:! type) {
+  fn Convert[self: Self]() -> T;
+}
+
+interface ImplicitAs(T:! type) {
+  fn Convert[self: Self]() -> T;
+}
+
+final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T) {
+  fn Convert[self: Self]() -> T = U.(ImplicitAs(T).Convert);
+}
+```
+
+Any other `impl` of `As(T)` is going to overlap unless we could establish that
+the type does not implement `ImplicitAs(T)`. However, in a generic context, we
+don't support (and don't intend to support) that sort of negative constraint. So
+there would be no way to prove the following `impl` of `As(T)` doesn't overlap
+the `final impl` of `As(T)`:
+
+```carbon
+import Core;
+
+class X(T:! type) {
+  impl as As(T);
+}
+```
+
+In particular, an unrelated library defining `MyType` could define
+`impl X(MyType) as ImplicitAs(MyType)`, which would create a conflict between
+the `impl` in `X(T)` and the `final impl` associated with `ImplicitAs(T)`.
+
+This motivates a change to the rules: we want to allow a non-final `impl` to
+overlap with a final `impl`. The question is what to do for queries in the
+overlap.
+
+Now consider this second example:
+
+```carbon
+import Core;
+
+class Array(Element:! type) { ... }
+
+// Impl 1
+impl forall [T:! type, U:! As(T)]
+    Array(U) as As(Array(T)) { ... }
+
+// Impl 2
+impl forall [T:! type, U:! ImplicitAs(T)]
+    Array(U) as ImplicitAs(Array(T)) { ... }
+```
+
+In this example, there is overlap between the definition of `As` provided by
+"Impl 1", and the definition of `As` provided by the `final impl` from
```

Comment:
Reference the definition of `package Core` from the beginning of this section and link to it `[Final `impl` priority](#final-impl-priority)`